File: leetcode/2487-remove-nodes-from-linked-list.js
// https://leetcode.com/problems/remove-nodes-from-linked-list/

// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var removeNodes = function(head) {
  const monotonicStack = [];

  while (head) {
    while (monotonicStack.length > 0 && monotonicStack[monotonicStack.length - 1] < head.val) {
      monotonicStack.pop();
    }
    monotonicStack.push(head.val);
    head = head.next;
  }

  const newHead = new ListNode(monotonicStack[0]);

  let curr = newHead;
  for (let i = 1; i < monotonicStack.length; i += 1) {
    curr.next = new ListNode(monotonicStack[i]);
    curr = curr.next;
  }

  return newHead;
};

File: leetcode/2487-remove-nodes-from-linked-list.py
# https://leetcode.com/problems/remove-nodes-from-linked-list/

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        monotonic_stack = []
        while head:
            while monotonic_stack and monotonic_stack[-1] < head.val:
                monotonic_stack.pop()
            monotonic_stack.append(head.val)
            head = head.next
        
        new_head = ListNode(monotonic_stack[0])
        
        curr = new_head
        for i in range(1, len(monotonic_stack), 1):
            curr.next = ListNode(monotonic_stack[i])
            curr = curr.next
        
        return new_head
